library/test/tests/hit-thread-limit.rs

```diff
@@ -0,0 +1,21 @@
+#![cfg(target_os = "linux")]
+#![feature(unwind_attributes)]
+
+use std::{io::ErrorKind, thread::Builder};
+
+#[unwind(aborts)]
+extern "C" fn set_thread_limit() {
+    assert_eq!(
+        unsafe { libc::setrlimit(libc::RLIMIT_NPROC, &libc::rlimit { rlim_cur: 0, rlim_max: 0 }) },
+        0
+    );
+}
+
+#[link_section = ".init_array"]
+#[used]
+static SET_THREAD_LIMIT: extern "C" fn() = set_thread_limit;
+
+#[test]
+fn thread_spawn_would_block() {
+    assert_eq!(Builder::new().spawn(|| unreachable!()).unwrap_err().kind(), ErrorKind::WouldBlock);
+}
```

Comment:
I am not a fan of the init_array here; it feels pretty unusual to me. Could we use a run-make test instead? You can likely find some examples in src/test/run-make; I'd expect us to have a simple test file, compile it with rustc --test, and then attempt to run the output with ulimit set. We'll want to be careful to not set the limit for more than that process.